working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+  - The desire for a zero cost wrapper type (motivated largely by interop
+   concerns), described briefly
+   [here](https://github.com/dart-lang/language/issues/1474)
+  - The desire to have compact syntax for so called "data classes" or "value
+types", discussed among other places
+[here](https://github.com/dart-lang/language/issues/314)
+  - The desire to have an accounting for data structures without identity as
+    described [here](https://github.com/dart-lang/language/issues/2246).
+
+The proposal here takes two steps.  The first is to add a restricted kind of
+class (provisionally describe as a "struct" here), which gives up some of the
+affordances of general Dart classes in exchange for compact syntax and automatic
+generation of useful methods.  These restricted classes provide data class like
+functionality.  The second step is to support a further restriction on structs
+with a single field which eliminates the wrapper object, representing the struct
+entirely as the underlying object (at the cost of making the abstraction
+entirely static).
+
+This proposal builds on previous proposals in this space, including:
+  - [Views](https://github.com/dart-lang/language/blob/master/working/1426-extension-types/feature-specification-views.md)
+  - [Extension types](https://github.com/dart-lang/language/blob/master/working/1426-extension-types/feature-specification.md)
+  - [Protected extension types](https://github.com/dart-lang/language/issues/1467)
+
+
+This overview is not intended to be a feature specification: it is designed to
+give a brief overview of the core proposal and set out some design points for
+discussion.  If there is buy-in on these ideas, we may choose to incorporate
+them into an existing proposal, or add a new feature specification.
+
+
+## Design principles
+
+We aim to minimize the differences between structs and classes.  As much as
+possible, structs should behave as restrictions of classes, and extension
+structs should behave as further restrictions of structs.  We specifically aim
+to avoid as much as possible having different behaviors for the same concept
+(e.g. differences in scoping).
+
+We also aim to minimize the amount of new syntax required, and to maximize the
+amount of new functionality that we can provide relative to the syntactic real
+estate consumed, and the new cognitive load imposed on users.
+
+## Roadmap
+
+The first section describes the full struct feature.  The second section
+describes the restriction of the full struct to support static wrappers.
+
+## Structs
+
+This section describes how structs look and behave, largely by example.  In many
+cases, alternatives or extensions to the core proposal are described in line.
+Larger extensions are postponed to a later section.
+
+## Structs by example
+
+Structs allow you to define simple classes very compactly.  For example, we
+might model a component vector as follows:
+
+```
+struct Component(int x, int y, int z, int w = 1);
```

Comment:
I like the compactness of the struct declaration, but to be honest I'm torn here.

Two thoughts come to mind:
* is there a more incremental way to introduce structs mapping to previous syntax?
* is the new syntax useful more generally beyond the concept of structs?

One incremental syntax I'd imagine is:

```dart
struct Component {
  int x;
  int y;
  int z;
  int w;

  Component(this.x, this.y, this.z, [this.w = 1]);
}
```

Another incremental step would be to make the constructor implicit (now instead of 0-args as we did with classes, we make struct constructors take all args).

Having a more incremental syntax like this, could potentially enable us to consider the new compact syntax as an alternative to redundant declaration/initialization in classes.  For example, we could make the two following programs equivalent:
```dart
class Foo(int x) extends Bar {
  foo() => x;
}
```

```dart
class Foo extends Bar {
  int x;
  Foo(this.x);
  foo() => x;
}
```

Reply:
> One incremental syntax I'd imagine is:

Yes, I considered that (and have added this to the alternatives section), though not with an explicit constructor - getting a default constructor is a big part of the deal here.   There are two things I don't like.  First, is that it feels very heavyweight relative to the terseness of the primary constructor.  I really do kind of feel that if we do this, we should get maximum bang for the buck, and I think the fact that you can literally just write a single line and get all the goodies is appealing.  Secondly, we need to require these to be final if we're going to use structural identity (I guess technically we could box each mutable field, but bleh), and I don't want to have to explicitly mark them as final.  But I'm also not super keen on `int x`  meaning one thing in one place and a different thing somewhere else.

> For example, we could make the two following programs equivalent:

Yeah, we could - I think  @munificent has a doc exploring that somewhere.

Reply:
> ... I really do kind of feel that if we do this, we should get maximum bang for the buck, and I think the fact that you can literally just write a single line and get all the goodies is appealing

Yeah, agree, that's very appealing.

Just to be clear: I'm not vouching for dropping all new syntax. Instead, I'd really like to have a layered approach. Just like UI-as-code: we'll recommend the new syntax, but we can use the desugared form to guide developers in how the new concepts work, highlight how we are mostly adding restrictions that can be leveraged for optimizations, and possibly reduce the implementation cost.

Reply:
Filed https://github.com/dart-lang/language/issues/2364 to discuss adding general primary constructors.